### PR TITLE
Add spread member completion item in list constructor context

### LINKED ISSUE #35507: Add Spread Operator support in Language Server
**Description:**
$subject.

### PR DESCRIPTION
**Pull request by malinthar**
## Purpose
$subject and refactor Spread field completion implementation

Fixes #35507 

## Samples
image

## Check List 
- [x] Read the [Contributing Guide](https://github.com/ballerina-platform/ballerina-lang/blob/master/CONTRIBUTING.md)
- [ ] Updated Change Log
- [ ] Checked Tooling Support (#<Issue Number>)
- [x] Added necessary tests
  - [x] Unit Tests
  - [ ] Spec Conformance Tests
  - [ ] Integration Tests
  - [ ] Ballerina By Example Tests
- [x] Increased Test Coverage   
- [ ] Added necessary documentation  
  - [ ] API documentation 
  - [ ] Module documentation in Module.md files
  - [ ] Ballerina By Examples

### CODE REVIEW

File: language-server/modules/langserver-core/src/test/resources/completion/expression_context/source/list_constructor_ctx_source6.bal [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+    int[] intArr6 = [...];
+}
+
+function getArray() returns int[] {
+
+}
+
+function getInt() returns int {
+
+}
```
Comment (KavinduZoysa) [inline language-server/modules/langserver-core/src/test/resources/completion/expression_context/source/list_constructor_ctx_source6.bal:32]:
```ballerina
function foo() returns int[]|error {
    return [1, 2];
}

public function main() returns error? {
    int[] arr = [<cursor>];
}
```
For this case we do not suggest `check` with `foo()`. Can we do that?

File: language-server/modules/langserver-core/src/test/resources/completion/expression_context/source/list_constructor_ctx_source6.bal [hunk 2]
```diff
@@ -0,0 +1,32 @@
+int[] arr = [1, 2, 3];
```
Comment (KavinduZoysa) [inline language-server/modules/langserver-core/src/test/resources/completion/expression_context/source/list_constructor_ctx_source6.bal:1]: Let's rename this file to show that we are doing `spead` related things.

File: language-server/modules/langserver-core/src/test/resources/completion/expression_context/source/list_constructor_ctx_source6.bal (hunk 1 above)
Comment (malinthar) [reply language-server/modules/langserver-core/src/test/resources/completion/expression_context/source/list_constructor_ctx_source6.bal:32]: You mean adding a completion item like `check foo()`

Comment (malinthar) [reply language-server/modules/langserver-core/src/test/resources/completion/expression_context/source/list_constructor_ctx_source6.bal:32]: Consider the following case 
image

We are not suggesting `check` with the expression in the above case as well. However, we suggest the `add check` code action to check the error.

image

In the case above we don't get the code action. Created https://github.com/ballerina-platform/ballerina-lang/issues/41165 to track this

File: language-server/modules/langserver-core/src/test/resources/completion/expression_context/source/list_constructor_ctx_source6.bal (hunk 2 above)
Comment (malinthar) [reply language-server/modules/langserver-core/src/test/resources/completion/expression_context/source/list_constructor_ctx_source6.bal:1]: Done

**Review by KavinduZoysa — DISMISSED** [review]

File: language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/completions/providers/context/ListConstructorExpressionNodeContext.java [hunk 3]
```diff
@@ -55,15 +70,69 @@ public List<LSCompletionItem> getCompletions(BallerinaCompletionContext context,
...
+                        }
+                        typeDescriptor = typeSymbol.get();
+                    } else {
+                        return false;
+                    }
+                    if (typeDescriptor.typeKind() != TypeDescKind.ARRAY) {
+                        return false;
+                    }
+                    return ((ArrayTypeSymbol) typeDescriptor)
+                            .memberTypeDescriptor().subtypeOf(expectedType.get());
```
Comment (mindula) [inline language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/completions/providers/context/ListConstructorExpressionNodeContext.java:118]:
```suggestion
                    return typeDescriptor.typeKind() == TypeDescKind.ARRAY && ((ArrayTypeSymbol) typeDescriptor).memberTypeDescriptor().subtypeOf(expectedType.get());

```

Comment (malinthar) [reply language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/completions/providers/context/ListConstructorExpressionNodeContext.java:118]: +1

File: language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/completions/SpreadCompletionItem.java [hunk 4]
```diff
@@ -0,0 +1,27 @@
+package org.ballerinalang.langserver.completions;
```
[the remark this answers is not in the document]
Comment (malinthar) [reply language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/completions/SpreadCompletionItem.java:16]: Done

**Review by mindula — APPROVED** [review]
LGTM

**Review by KavinduZoysa — APPROVED** [review]

**Pull request merged**